modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1994,4 +2070,252 @@ private void processDelay(InetSocketAddress sockAddr) {
             reconnectionDelaysRemainder.put(sockAddr, delay);
         }
     }
+
+    /**
+     * JDBC implementation of {@link MarshallerContext}.
+     */
+    private class JdbcMarshallerContext extends BlockingJdbcChannel implements MarshallerContext {
+        /** Type ID -> class name map. */
+        private Map<Integer, String> cache = new ConcurrentHashMap<>();
+
+        /** {@inheritDoc} */
+        @Override public boolean registerClassName(
+            byte platformId,
+            int typeId,
+            String clsName,
+            boolean failIfUnregistered
+        ) throws IgniteCheckedException {
+            assert platformId == MarshallerPlatformIds.JAVA_ID
+                : String.format("Only Java platform is supported [expPlatformId=%d, actualPlatformId=%d].", MarshallerPlatformIds.JAVA_ID, platformId);
+
+            boolean res = true;
+            if (!cache.containsKey(typeId)) {
+                try {
+                    JdbcUpdateBinarySchemaResult updateRes = doRequest(new JdbcBinaryTypeNamePutRequest(typeId, platformId, clsName));
+                    res = updateRes.success();
+                }
+                catch (ExecutionException | InterruptedException | ClientException | SQLException e) {
+                    throw new IgniteCheckedException(e);
+                }
+
+                if (res)
```

Comment:
Unsuccessful result means that the new type was not registered by grid. I think there should be no special handling because BinaryContext already know how to deal with it.